xla/service/gpu/runtime/nccl_collective_permute_thunk.cc

```diff
@@ -53,6 +53,31 @@ absl::StatusOr<const int64_t> GetCurrentId(
           : current_logical_id.computation_id;
   return current_id;
 }
+
+bool IsLocalPeerTransfer(
+    const NcclP2PConfig::SourceTargetMapEntry& source_target,
+    se::Stream& stream, const int64_t current_id, const int64_t device_count) {
+  const std::optional<int64_t> source_id = source_target.source;
+  const std::optional<int64_t> target_id = source_target.target;
+  // Since mixing nccl p2p with p2p memcopy will cause random deadlocks.
+  // We determine if it's a local peer by the following conditions:
+  // 1. Both source and target IDs are present and they are within a node
+  // 2. Source ID is present, but target ID is not.
```

Comment:
What does it mean if the source ID is present, but target ID is not? You can effectively not tell if this is within a host, can you?

Reply:
it means this rank is only receiving from a peer but not sending data to anybody. We can still use the source id and current id to determine that this is a local peer or not. Since there's no destination we need to send, it's still a local copy.

Reply:
sure updated.